src/lua/datetime.lua

```diff
@@ -420,6 +625,221 @@ local function datetime_tostring(self)
+
+    -- left is date, right is interval
+    if not left_interval and right_interval then
+        year, month, secs, nsec, adjust = interval_decouple_args(rhs)
+        return datetime_increment_by(datetime_new_copy(lhs), -1, year, month,
+                                     secs, nsec, adjust)
+    -- left is date, right is date
+    elseif not left_interval and not right_interval then
+        local obj = interval_new()
+        obj.sec, obj.nsec = normalize_nsec(lhs.epoch - rhs.epoch,
+                                           lhs.nsec - rhs.nsec)
+        return obj
+    -- both left and right are intervals
+    elseif left_interval and right_interval then
+        year, month, secs, nsec, adjust = interval_decouple_args(lhs)
+        local obj = interval_init(year, month, secs, nsec, adjust)
+        year, month, secs, nsec = interval_decouple_args(rhs)
+        obj.year = obj.year - year
+        obj.month = obj.month - month
+        obj.sec, obj.nsec = normalize_nsec(obj.sec - secs, obj.nsec - nsec)
+        return obj
+    else
+        error_incompatible("operator -")
+    end
+end
+
+--[[
+Matrix of addition operands eligibility and their result type
+
+|                 | datetime | interval | table    |
++-----------------+----------+----------+----------+
+| datetime        |          | datetime | datetime |
+| interval        | datetime | interval | interval |
+| table           |          |          |          |
+]]
+local function datetime_interval_add(lhs, rhs)
+    lhs, rhs = date_first(lhs, rhs)
+
+    check_date_interval(lhs, "operator +")
+    check_interval_table(rhs, "operator +")
+    local left_interval = is_table(lhs) or is_interval(lhs)
+    local right_interval = is_table(rhs) or is_interval(rhs)
+    local year, month, secs, nsec, adjust
+
+    -- left is date, right is interval
+    if not left_interval and right_interval then
+        local obj = datetime_new_copy(lhs)
+        year, month, secs, nsec, adjust = interval_decouple_args(rhs)
+        return datetime_increment_by(obj, 1, year, month, secs, nsec, adjust)
+    -- both left and right are intervals
+    elseif left_interval and right_interval then
+        year, month, secs, nsec, adjust = interval_decouple_args(lhs)
+        local obj = interval_init(year, month, secs, nsec, adjust)
+        year, month, secs, nsec = interval_decouple_args(rhs)
+        obj.year = obj.year + year
+        obj.month = obj.month + month
+        obj.sec, obj.nsec = normalize_nsec(obj.sec + secs, obj.nsec + nsec)
+        return obj
+    else
+        error_incompatible("operator +")
+    end
+end
+
+local function interval_cmp(lhs, rhs, raising)
+    if not is_interval(lhs) or not is_interval(rhs) then
+        if raising then
+            error('incompatible types for interval comparison', 3)
+        else
+            return nil
+        end
+    end
+    local sdiff = lhs.sec - rhs.sec
+    return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
+end
```

Comment:
```
tarantool> i1 = interval.new({year = 1})
---
...

tarantool> i2 = interval.new({year = 2})
---
...

tarantool> i1 == i2
---
- true
...
-- Ensure they are not empty.
tarantool> i1
---
- +1 years
...

tarantool> i2
---
- +2 years
...
```
Seems not to be working correctly. I doubt you can compare them at all. One of them might have months, other - seconds. The first one's value in seconds depends on the date it is applied to. So it can be both `>` and `<` and `==` the second interval.

Reply:
That's very good question, thanks! On one hand we could hardly compare them precisely, because their final values are date-dependent. But on the other hand we could apply some strict hierarchical rules (which were already defined in RFC for interval arithmetic, when we always deal in particular order, i.e. year > month > week > day > etc). 

For our case it would be: difference in years always more important than differences in months, which, in turn, would always be more important than any difference of seconds.

```
diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index 6c5f996f1..d68cc30c9 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -822,6 +822,14 @@ local function interval_cmp(lhs, rhs, raising)
             return nil
         end
     end
+    local ydiff = lhs.year - rhs.year
+    if ydiff ~= 0 then
+        return ydiff
+    end
+    local mdiff = lhs.month - rhs.month
+    if mdiff ~= 0 then
+        return mdiff
+    end
     local sdiff = lhs.sec - rhs.sec
     return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
 end
```

This would provide reasonable results if year or month attributes involved:
```
tarantool> inew = date.interval.new
---
...

tarantool> I1 = inew{year = 1}
---
...

tarantool> I2 = inew{year = 2}
---
...

tarantool> I1 == I2
---
- false
...

tarantool> I1 < I2
---
- true
...
```

OTOH human knows that 380 days always larger than 1 year, but for such hierarchical algorithm it would be quite contrary:
```
tarantool> I3 = inew{days = 380}
---
...

tarantool> I1 > I3
---
- true
...

tarantool> I1 == I3
---
- false
...
```

But it's not a big problem if we have consistent rules for comparison.

Reply:
And adding test to check this ordered comparison:
```diff
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -577,7 +577,7 @@ test:test("Time interval tostring()", function(test)
 end)
 
 test:test("Time interval comparisons", function(test)
-    test:plan(36)
+    test:plan(66)
     -- check empty arguments
     local int1 = date.interval.new()
     test:is(int1.sec, 0, "ts.sec == 0")
@@ -625,6 +625,46 @@ test:test("Time interval comparisons", function(test)
     assert_raises(test, incompat_int_types, function() return int1 >= nil end)
     assert_raises(test, incompat_int_types, function() return int1 >= 123 end)
     assert_raises(test, incompat_int_types, function() return int1 >= '1970-01-01' end)
+
+    local compare_intervals = {
+        {year = -2, month = -1, sec = -1e6, nsec = -2e5},
+        {year = -2, month = -1, sec = -1e6, nsec = -1e5},
+        {year = -2},
+        {year = -1},
+        {month = -20},
+        {month = -10},
+        {month = -1},
+        {day = -400},
+        {day = -256},
+        {day = -30},
+        {sec = -2},
+        {nsec = -1e9},
+        {nsec = -1e6},
+        {nsec = -1e3},
+        {nsec = -1},
+        {nsec = 1},
+        {nsec = 1e3},
+        {nsec = 1e6},
+        {nsec = 1e9},
+        {sec = 1, nsec = 1},
+        {day = 1},
+        {day = 30},
+        {day = 256},
+        {day = 400},
+        {month = 1},
+        {month = 10},
+        {month = 20},
+        {year = 1},
+        {year = 2},
+        {year = 2, month = 1, sec = 1e6, nsec = 1e5},
+        {year = 2, month = 1, sec = 1e6, nsec = 2e5},
+    }
+
+    for i = 1, #compare_intervals - 1 do
+        local before = date.interval.new(compare_intervals[i])
+        local after = date.interval.new(compare_intervals[i + 1])
+        test:ok(before < after, ('%s < %s'):format(before, after))
+    end
 end) 
```

Reply:
We completely need no comparisons of intervals. They are mostly undefined behaviour.

Reply:
Please, completely ban comparisons, to avoid unambiguity.

Reply:
Sigh. Removed.

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -829,30 +829,6 @@ local function datetime_interval_add(lhs, rhs)
     end
 end
 
-local function interval_cmp(lhs, rhs, raising)
-    if not is_interval(lhs) or not is_interval(rhs) then
-        if raising then
-            error('incompatible types for interval comparison', 3)
-        else
-            return nil
-        end
-    end
-    local sdiff = lhs.sec - rhs.sec
-    return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
-end
-
-local function interval_eq(lhs, rhs)
-    return interval_cmp(lhs, rhs, false) == 0
-end
-
-local function interval_lt(lhs, rhs)
-    return interval_cmp(lhs, rhs, true) < 0
-end
-
-local function interval_le(lhs, rhs)
-    return interval_cmp(lhs, rhs, true) <= 0
-end
-
 --[[
     Create datetime object representing current time using microseconds
     platform timer and local timezone information.
@@ -1166,9 +1142,6 @@ end
 
 ffi.metatype(interval_t, {
     __tostring = interval_tostring,
-    __eq = interval_eq,
-    __lt = interval_lt,
-    __le = interval_le,
     __sub = datetime_interval_sub,
     __add = datetime_interval_add,
     __index = interval_index,

```

Reply:
Also interval comparison checks removed:

```diff
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -5,7 +5,7 @@ local test = tap.test('errno')
 local date = require('datetime')
 local ffi = require('ffi')
 
-test:plan(23)
+test:plan(22)
 
 -- minimum supported date - -5879610-06-22
 local MIN_DATE_YEAR = -5879610
@@ -28,7 +28,6 @@ local MAX_MIN_RANGE = MAX_HOUR_RANGE * 60
 local MAX_SEC_RANGE = MAX_DAY_RANGE * SECS_PER_DAY
 
 local incompat_types = 'incompatible types for datetime comparison'
-local incompat_int_types = 'incompatible types for interval comparison'
 local only_integer_ts = 'only integer values allowed in timestamp'..
                         ' if nsec, usec, or msecs provided'
 local only_one_of = 'only one of nsec, usec or msecs may be defined'..
@@ -577,57 +576,6 @@ test:test("Time interval tostring()", function(test)
     end
 end)
 
-test:test("Time interval comparisons", function(test)
-    test:plan(36)
-    -- check empty arguments
-    local int1 = date.interval.new()
-    test:is(int1.sec, 0, "ts.sec == 0")
-    test:is(int1.nsec, 0, "ts.nsec == 0")
-    test:is(int1.year, 0, "ts.year == 0")
-    test:is(int1.month, 0, "ts.month == 0")
-    test:is(tostring(int1), "0 seconds", "tostring(int1)")
-    -- check empty table
-    local int2 = date.interval.new{}
-    test:is(int2.sec, 0, "ts.sec ==0")
-    test:is(int2.nsec, 0, "ts.nsec == 0")
-    test:is(int2.year, 0, "ts.year == 0")
-    test:is(int2.month, 0, "ts.month == 0")
-    test:is(tostring(int2), "0 seconds", "tostring(int2)")
-    -- check their equivalence
-    test:is(int1, int2, "int1 == int2")
-    test:is(int1 ~= int2, false, "not int1 != int2")
-
-    test:isnt(int1, nil, "int1 != {}")
-    test:isnt(int1, {}, "int1 != {}")
-    test:isnt(int1, "1970-01-01T00:00:00Z", "int1 != '1970-01-01T00:00:00Z'")
-    test:isnt(int1, 19700101, "int1 != 19700101")
-
-    test:isnt(nil, int1, "{} ~= int1")
-    test:isnt({}, int1 ,"{} ~= int1")
-    test:isnt("1970-01-01T00:00:00Z", int1, "'1970-01-01T00:00' ~= int1")
-    test:isnt(19700101, int1, "int1 ~= int1")
-
-    test:is(int1 < int2, false, "not int1 < int2")
-    test:is(int1 > int2, false, "not int1 < int2")
-    test:is(int1 <= int2, true, "not int1 < int2")
-    test:is(int1 >= int2, true, "not int1 < int2")
-
-    -- check comparison errors -- ==, ~= not raise errors, other
-    -- comparison operators should raise
-    assert_raises(test, incompat_int_types, function() return int1 < nil end)
-    assert_raises(test, incompat_int_types, function() return int1 < 123 end)
-    assert_raises(test, incompat_int_types, function() return int1 < '1970-01-01' end)
-    assert_raises(test, incompat_int_types, function() return int1 <= nil end)
-    assert_raises(test, incompat_int_types, function() return int1 <= 123 end)
-    assert_raises(test, incompat_int_types, function() return int1 <= '1970-01-01' end)
-    assert_raises(test, incompat_int_types, function() return int1 > nil end)
-    assert_raises(test, incompat_int_types, function() return int1 > 123 end)
-    assert_raises(test, incompat_int_types, function() return int1 > '1970-01-01' end)
-    assert_raises(test, incompat_int_types, function() return int1 >= nil end)
-    assert_raises(test, incompat_int_types, function() return int1 >= 123 end)
-    assert_raises(test, incompat_int_types, function() return int1 >= '1970-01-01' end)
-end)
-
 test:test("Time interval __index fields", function(test)
     test:plan(11)
 
```